{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Project 3 - TLR 2 vs TLR 4 & TLR 6: Who is the imposter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Intoduction\n",
    "\n",
    "Toll-like receptors (TLRs) are a class of proteins that play a key role in the innate immune system. They are single-pass membrane-spanning receptors usually expressed on sentinel cells such as macrophages and dendritic cells, that recognize structurally conserved molecules derived from microbes. Once these microbes have breached physical barriers such as the skin or intestinal tract mucosa, they are recognized by TLRs, which activate immune cell responses. The TLRs include TLR1, TLR2, TLR3, TLR4, TLR5, TLR6, TLR7, TLR8, TLR9, TLR10, TLR11, TLR12, and TLR13, though the last three are not found in humans. All three of the TLRs we will focus on in this project, ie TLR2, TLR4, and TLR6 are located on the cell membrane and posess three distinct domains - The extracellular domain, The transmembrane domain, and the intracellular domain.\n",
    "\n",
    "The ability of the immune system to recognize molecules that are broadly shared by pathogens is, in part, due to the presence of immune receptors like toll-like receptors.\n",
    "\n",
    "The binding of ligands - either in the form of adjuvant used in vaccinations or in the form of invasive moieties during times of natural infection - to the TLR marks the key molecular events that ultimately lead to innate immune responses and the development of antigen-specific acquired immunity.\n",
    "\n",
    "Upon activation, TLRs recruit adaptor proteins (proteins that mediate other protein-protein interactions) within the cytosol of the immune cell in order to propagate the antigen-induced signal transduction pathway. These recruited proteins are then responsible for the subsequent activation of other downstream proteins and ultimately the transcription of cytokines.\n",
    "\n",
    "TLR2 is a major mammalian TLR that can recognize lipoproteins derived from bacteria, viruses, fungi, and parasites. \n",
    "\n",
    "TLR4 has been long recognized as the sensing receptor for gram-negative lipopolysaccharide (LPS). \n",
    "\n",
    "TLR 6 recognizes ligands such multiple diacyl lipopeptides derived from gram-positive bacteria and mycoplasma and several fungal cell wall saccharides. TLR 6 dimerizes with TLR2 for activation of downstream signalling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scientific Question\n",
    "How similar are pathogen receptors from receptors that recognize different pathogenic ligands and from receptors they evolved from. I want to study this especially in relation to human TLR 2 in comparison to human TLR 4 and human TLR 6. I selected these genes because TLR2, TLR4, and TLR6 recognize motifs from similar bacteria. TLR4 evolved from TLR2 whereas TLR 6 dimerizes with TLR 2 and might share similar regions like transmembrane domains. Since TLR2 has different yet direct relationships with TLR4 and TLR6, it will be interesting to study the genetic similarities between them"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hypothesis\n",
    "Given the relationship that exists between TLR2 and TLR 4 as well as TLR2 and TLR6, I hypothesize that TLR2 and TLR4 share a higher similarity in their extracellular domain that is responsible for binding ligands as compared to TLR 2 and TLR 6 whereas TLR 2 and TLR 6 share a higher similarity in their transmembrane and intracellular domain as these regions interact via dimerization to activate downstream signaling. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Methods"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To study the relationship between TLR2, TLR4, and TLR 6, it is important to divide each gene in its 3 structural domains - The extracellular domain, the transmembrane domain, and the cytoplasmic domain. In this project, we will be studying the genetic similarities between these three regions across TLR2, TLR4 and TLR6. To do so, we will be using pairwise sequence alignment. The fasta files for the protein sequences were obtained from uniprot\n",
    "\n",
    "We will then be representing the data using grouped parplots and drawing conclusions on the same.\n",
    "\n",
    "For the increased understanding of the viewer, we will also be representing the protein structures of the TLR2 and TLR4 protein and explaining the protein motifs present. The ascension numbers for the proteins were obtained from rcsb.org."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Importing "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: nglview in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (2.7.7)\n",
      "Requirement already satisfied: numpy in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from nglview) (1.16.2)\n",
      "Requirement already satisfied: ipywidgets>=7 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from nglview) (7.4.2)\n",
      "Requirement already satisfied: ipykernel>=4.5.1 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from ipywidgets>=7->nglview) (5.1.0)\n",
      "Requirement already satisfied: widgetsnbextension~=3.4.0 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from ipywidgets>=7->nglview) (3.4.2)\n",
      "Requirement already satisfied: nbformat>=4.2.0 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from ipywidgets>=7->nglview) (4.4.0)\n",
      "Requirement already satisfied: ipython>=4.0.0; python_version >= \"3.3\" in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from ipywidgets>=7->nglview) (7.4.0)\n",
      "Requirement already satisfied: traitlets>=4.3.1 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from ipywidgets>=7->nglview) (4.3.2)\n",
      "Requirement already satisfied: jupyter-client in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from ipykernel>=4.5.1->ipywidgets>=7->nglview) (5.2.4)\n",
      "Requirement already satisfied: tornado>=4.2 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from ipykernel>=4.5.1->ipywidgets>=7->nglview) (6.0.2)\n",
      "Requirement already satisfied: notebook>=4.4.1 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from widgetsnbextension~=3.4.0->ipywidgets>=7->nglview) (5.7.8)\n",
      "Requirement already satisfied: ipython-genutils in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from nbformat>=4.2.0->ipywidgets>=7->nglview) (0.2.0)\n",
      "Requirement already satisfied: jsonschema!=2.5.0,>=2.4 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from nbformat>=4.2.0->ipywidgets>=7->nglview) (3.0.1)\n",
      "Requirement already satisfied: jupyter-core in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from nbformat>=4.2.0->ipywidgets>=7->nglview) (4.4.0)\n",
      "Requirement already satisfied: decorator in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets>=7->nglview) (4.4.0)\n",
      "Requirement already satisfied: prompt-toolkit<2.1.0,>=2.0.0 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets>=7->nglview) (2.0.9)\n",
      "Requirement already satisfied: setuptools>=18.5 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets>=7->nglview) (40.8.0)\n",
      "Requirement already satisfied: backcall in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets>=7->nglview) (0.1.0)\n",
      "Requirement already satisfied: pygments in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets>=7->nglview) (2.3.1)\n",
      "Requirement already satisfied: colorama; sys_platform == \"win32\" in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets>=7->nglview) (0.4.1)\n",
      "Requirement already satisfied: pickleshare in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets>=7->nglview) (0.7.5)\n",
      "Requirement already satisfied: jedi>=0.10 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets>=7->nglview) (0.13.3)\n",
      "Requirement already satisfied: six in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from traitlets>=4.3.1->ipywidgets>=7->nglview) (1.12.0)\n",
      "Requirement already satisfied: python-dateutil>=2.1 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from jupyter-client->ipykernel>=4.5.1->ipywidgets>=7->nglview) (2.8.0)\n",
      "Requirement already satisfied: pyzmq>=13 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from jupyter-client->ipykernel>=4.5.1->ipywidgets>=7->nglview) (18.0.0)\n",
      "Requirement already satisfied: nbconvert in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.4.0->ipywidgets>=7->nglview) (5.4.1)\n",
      "Requirement already satisfied: terminado>=0.8.1 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.4.0->ipywidgets>=7->nglview) (0.8.1)\n",
      "Requirement already satisfied: Send2Trash in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.4.0->ipywidgets>=7->nglview) (1.5.0)\n",
      "Requirement already satisfied: jinja2 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.4.0->ipywidgets>=7->nglview) (2.10)\n",
      "Requirement already satisfied: prometheus-client in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.4.0->ipywidgets>=7->nglview) (0.6.0)\n",
      "Requirement already satisfied: attrs>=17.4.0 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from jsonschema!=2.5.0,>=2.4->nbformat>=4.2.0->ipywidgets>=7->nglview) (19.1.0)\n",
      "Requirement already satisfied: pyrsistent>=0.14.0 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from jsonschema!=2.5.0,>=2.4->nbformat>=4.2.0->ipywidgets>=7->nglview) (0.14.11)\n",
      "Requirement already satisfied: wcwidth in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from prompt-toolkit<2.1.0,>=2.0.0->ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets>=7->nglview) (0.1.7)\n",
      "Requirement already satisfied: parso>=0.3.0 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from jedi>=0.10->ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets>=7->nglview) (0.3.4)\n",
      "Requirement already satisfied: mistune>=0.8.1 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.4.0->ipywidgets>=7->nglview) (0.8.4)\n",
      "Requirement already satisfied: entrypoints>=0.2.2 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.4.0->ipywidgets>=7->nglview) (0.3)\n",
      "Requirement already satisfied: bleach in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.4.0->ipywidgets>=7->nglview) (3.1.0)\n",
      "Requirement already satisfied: pandocfilters>=1.4.1 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.4.0->ipywidgets>=7->nglview) (1.4.2)\n",
      "Requirement already satisfied: testpath in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.4.0->ipywidgets>=7->nglview) (0.4.2)\n",
      "Requirement already satisfied: defusedxml in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.4.0->ipywidgets>=7->nglview) (0.5.0)\n",
      "Requirement already satisfied: MarkupSafe>=0.23 in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from jinja2->notebook>=4.4.1->widgetsnbextension~=3.4.0->ipywidgets>=7->nglview) (1.1.1)\n",
      "Requirement already satisfied: webencodings in c:\\users\\yashaswat\\anaconda3\\lib\\site-packages (from bleach->nbconvert->notebook>=4.4.1->widgetsnbextension~=3.4.0->ipywidgets>=7->nglview) (0.5.1)\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import SeqIO\n",
    "import matplotlib.pyplot as plot\n",
    "\n",
    "import nglview\n",
    "import numpy as np\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Defining the packages and explaining how they have been used\n",
    "\n",
    "Bio - Biopython is a set of freely available tools for biological computation written in Python. It is a distributed collaborative effort to develop Python libraries and applications which address the needs of current and future work in bioinformatics. Here, we will use Biopython to conduct a pairwise sequence alignment using pairwise2. format_alignment is a function within Bio.pairwise which gives a nice printout. Bio.SeqIO provides a simple uniform interface to input and output assorted sequence file formats. Here it will be used to read in fasta files\n",
    "\n",
    "numpy - NumPy is one of the most widely used python libraries in data science. It provides high-performance, easy to use structures and data analysis tools. It offers comprehensive mathematical functions, random number generators, linear algebra routines, Fourier transforms, and more. In this project, we will be using Numpy to create a data array of our sequence alignments score to be graphed \n",
    "\n",
    "matplotlib - Matplotlib is a comprehensive library for creating static, animated, and interactive visualizations in Python. We use matplotlib in this project to vizualize the differences in the pairwise sequence alignment score\n",
    "\n",
    "nglview - NGL Viewer is a collection of tools for web-based molecular graphics. Here we will use nglview to display the protein structure of TLR2 and TLR4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Explaination of Bioinformatics method used "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Vizualizing the protein \n",
    "The TLR2, TLR4 and TLR6 proteins will be vizualized in 3D utilized using nglview. This method takes in as its input a string containing the protein database id for our gene of interest. As its output, it gives a color coded protein with visual surface protein modifications."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pairwise Sequence Alignment "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pairwise Sequence Alignment is used to identify regions of similarity that may indicate functional, structural and/or evolutionary relationships between two biological sequences (here a potein sequence). In this project we will be using Biopython to perform pairwise sequence alignments on our different regions of interest. The code takes as its input a fasta file containg the protein sequence of iterest. The output is the print display of the sequence alignment as well as an alignmnet score. The higher the alignment score between 2 genes, the greater the structural and functional similarity.\n",
    "\n",
    "The alignment will be performed using the global alignments parameter and the scoring is as follows: Identical characters are given 2 points, 1 point is deducted for each non-identical character. Gaps are not penalized."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Vizualizing the TLR2 and TLR4 receptor protein's structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### TLR2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f192a4f3766c4ec8b1653790087faada",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "view_TLR2 = nglview.show_pdbid('6NIG') \n",
    "#assigning the TLR2 protein structure to view_TLR2 using TLR2's protein ID\n",
    "view_TLR2.add_cartoon(selection=\"protein\")\n",
    "view_TLR2.add_surface(selection=\"protein\", opacity=0.1)\n",
    "view_TLR2.render_image()\n",
    "#adding the correct color scheme to view_TLR2 and creating protein textures on it\n",
    "view_TLR2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Among the various TLRs, TLR2 has a special place due to its ability to sense the widest repertoire of PAMPs owing to its heterodimerization with either TLR1 or TLR6, broadening its ligand diversity against pathogens. \n",
    "In the protein structure above, the protein regions shaped like hooks and colored red and orange are the two arms of the dimerized receptor, one from TLR2(red) and one from TLR1/6. The lower end of the arms contains the C-terminus whereas the curved upper end contains the N-terminus."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### TLR 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "35ae02a7bb834caf9121f444246eb8e0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "view_TLR4 = nglview.show_pdbid('4G8A')  \n",
    "#assigning the TLR4 protein structure to view_TLR4 using TLR4's protein ID\n",
    "view_TLR4.add_cartoon(selection=\"protein\")\n",
    "view_TLR4.add_surface(selection=\"protein\", opacity=0.1)\n",
    "#adding the correct color scheme to view_TLR4 and creating protein textures on it\n",
    "view_TLR4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TLR4 is a type I transmembrane protein composed of 22 extracellular leucine-rich repeats (LRRs), a transmembrane domain, and the Toll/IL-1 receptor domain that is essential for TLR signaling and conserved among members of the Toll receptor family. TLR4 alone does not directly bind LPS and requires the coreceptor MD-2. MD-2 is associated with the extracellular domain of TLR4 and is indispensable for LPS recognition\n",
    "\n",
    "Like TLR2, the protein regions shaped like hooks and colored red and orange are the two arms of the TLR4 receptor. The lower end of the arm contains the C-terminus whereas the curved upper end contains the N-terminus. The light yellow and light orange protein regions between the two curved arms represent the MD-2 co-receptors. This curved region inside the arm contains the ligand binding domain"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reading in the fasta files for the extracellular domain of TLR 2, TLR 4 and TLR 6.\n",
    "This region contains the ligand binding region for all 3 TLRs. Apon ligand recognition, The TLRs activate downstream signalling.\n",
    "As previously mentioned, we hypothesized that for this region, TLR2 will be more similar to TLR4 as compared to TLR6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "TLR2EC = SeqIO.read('TLR2EC.fasta','fasta')\n",
    "\n",
    "TLR4EC = SeqIO.read('TLR4EC.fasta','fasta')\n",
    "\n",
    "TLR6EC = SeqIO.read('TLR6EC.fasta','fasta')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Performing pairwise sequence alignments: TLR 2 and TLR 4 vs TLR 2 and TLR 6 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ESSNQASLS---CDR-----NGI---CKGSSGS--LNS---IPSG--L---TEAVKS-LDLSN-N--RIT----YISN--SD---LQRCVNLQALV-LTSNG---IN-TIEEDSFSSLG---SLE-HLDLSYNY-LSN-LSSSWFK---PLS--SLTF-LN-----L--LGNPY-KTLG---ET---SL--FS---HLTK-LQI--LR-VG--NMDTFTK-IQR--KD----FAG--LTF-LEE-LEIDASDLQSYEPKSL-KSIQNVSHLI----L---HMKQHI--LL-LEIFVDVTSSVECLELRDTDLDTFHFSELSTGETNSL----I-----K-----KF-TF-RNVKIT--DESLFQ-VMKLLNQ--IS-GL--LE---L---EFDDCTL-NGV-GN---FRASDNDRVIDPGKV---ET-L---TIR---RLHIPRF-YLF-Y--D----LST---LY---S---L---T-ERVKRITVEN--SKV--F----L--VP-CLLS---QH---LKSLEY--LD--L---SENLMVEEYLK--NSAC--E-DAW-PSLQT--LI-L-RQNH-LASLE-KTGE--------TLLT-LKN-LTNIDI-SK-N-----SFHSMP-----ETCQWP-EKMKY-L-----NL---S--ST-----R--IH----SV-T-------GCIPKT---LEI--LD-VSNNNL----NLFSL---N-LPQLK----ELYISRNKLMT-LPDA-S---L--L-PM-----L--L-VLKI--SR-NAITT-F-S------KE-----Q-LD-SF--H--TL-KTL--E--------AGG--N---N-FI-CS-CEFL--SFT-QE----Q-QALAKVLIDWPANYL-------CDS--PSHVRGQQVQD-----VRLSVSECHR---T-----\n",
      "|       |   |       | |   |        ||    ||    |   |   |  ||||  |  |      | |   |    ||  |    |  | |     |  || ||     |   ||  |  ||    |   |         |    ||   |      |  |     | |    ||   ||  |    || | |    |  |   |       ||   |     |    ||  | | |     || |    |  | ||  |  |    |   |  |    || |        |     |   ||     | |     |      |     |     |  |  ||      | ||   |||      |  ||  ||   |   ||      |   ||   |         |  |    |  |   ||    ||      ||  |  |    |     |    |   |   | ||||        |    |    |  |  |      |    || |    |   |   | |       |  | |   | |   |||    |  | | |  | |   | |         |  | ||  |   |  |  |     |  |       |  |   |     |     ||   |  |      |  |     |  |       | |      |    |  |    |    |  |    | ||       ||   || | | | |  |   |  | |      |  | ||    |  |     | |      |      | || |   |  |  |    |        |    |   | |  |  ||    ||  |     | |     |       |       |    ||        |     | |  |       |     \n",
      "E-------SWEPC--VEVVPN-ITYQC------MELN-FYKIP--DNLPFST---K-NLDLS-FNPLR--HLGSY-S-FFS-FPELQ--V----L-DL-S--RCEI-QTI-ED-----GAYQSL-SH--LS---TL--IL------TGNP--IQSL--AL-GAFSGLSSL----QK-L-VAVETNLASLENF-PIGHL-KTL--KEL-NV-AHN------LIQ-SFK-LPEYF--SNLT-NL-EHL-----DL-S----S-NK-IQ--S--IYCTDLRVLH--Q--MPLLNL--------S-----L---DL-----S-L-----N--PMNFIQPGAFKEIRLHK-LT-LRN----NFD-SL--NVMK----TCI-QGLAGLEVHRLVLGEF-----RN--EGNLEKF---------D--K-SALE-GLCNLTI-EEFRL-----AYL-DYYLDDIIDL--FNCL-TNVSSFSLVSVTIERVK------DFS--YNFGWQHLELV-NC---KFGQ-FPTLK-L--KSL-KRLTFTS-N-------KGGN-A-FSEVD--LPSL--EFL-DLSR-N-GL-S--FK-G-CCSQSDFGT--TSLK-YL---D-LS-FNGVITMS--S--NFLGLE--Q--LE----HLDFQHSNLKQMSEFS-VFLSLRNLI-YLDIS-HTHTRVAFNG-I---FNGL--SSL-EV----LKMAGN--S-FQENFLP---DIFTEL---RN-L-TFL-D-LSQCQLEQLSP-TAFNSLSSLQVL--NMS-HN----NFFSLDTFPYK-CLNSLQVLDYS-LNHIMT-SK--KQELQHFPSSLA--FLNLTQNDF-AC-TCE--HQSF-LQ-WIKDQRQ-----L-------LVEVERMEC--ATPS--------DKQGMPV-L--S----LNITCQMNK\n",
      "  Score=250\n",
      "\n"
     ]
    }
   ],
   "source": [
    "alignments = pairwise2.align.globalxx(TLR2EC.seq, TLR4EC.seq)\n",
    "#The pairwise alignment method is called by pairwise2.align and the scoring parameters are stored in globalxx. \n",
    "#The input within the alignment function is the protein sequences in our two fasta files of interest.\n",
    "#All the different alignment functions are contained in the object 'align'.\n",
    "print(format_alignment(*alignments[0]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ESSNQ--ASLSC-DRNGICKGSS--GSLNSI--PSG--L---TEAVKS-LDL-SN-NRITYISNSDLQRCVNLQA-L-VLTSNGINTIEED-SFSSLGSLEHLD--LSYNY-LS-N---LSS---SW-FKPLSSLTFLNL--LGNP-YKTLGETS-LFSHLTK--LQ-ILR-----VGNMDT-FTKIQRK--DFAGLT-FLEELEI-D--ASDLQSYEP--KS----LKSIQ-NVSH-LI-LH--MK-QHI-L-LLE-IFVDVTS-----SVEC--LELRDTDLDTFHFS------ELSTG-ETN-SL-IK---KFTFRN---VKI---TDESLF--QVMKLLNQISG---L--LE-LEFDDCTLNGVG---NFRASDNDR---V-IDPGKV---ETLTIRR---L------HIP--------R-F---------YLF--YD-L----S------TLYS---L---T-ERVKR-ITVENSK-VFLVPCLL-SQHLKS--LEYLDL---SE-NL-MVEEY-LKN--SACED----------AWPSLQTLI---LR---QNHLA----SL--EKTGE--TL--L-TL---KN-L---------TNI-D-----I---SK-NSFH--SMP----ET-CQ-WP-EKMKY----LNLSS---TRIH-SVTG--CIPKTLE---ILD-VSNNNLN-LF-SLNLPQLKELYI-SRN--K----LMT--LPD-----A--SL--LPML-------L-VLKISR---N------AITT-F--SKE-QLD---SFHTL-KTLEAGGN-N-FI-CS-CEFLS--FT-----Q---EQQALAKVLID--WPAN--YL-CDS-P-SHV-RGQQVQDVRLSV-------SE--CHR--T-\n",
      "   |   |     |     |  |  | |  |  |    |   |   |  ||  |  |   ||            | | |  |        | ||  | | | |   |     || |   |     |  ||      | |   |    |  |     | ||     || |       |      |    |   |   |  |       |  |  |    |  |     | | | |    |  |   || |   | ||  |           |     | |   ||            |     ||  || |    | |      |     |  |||  ||    | ||    |  |  |     | |      |    | |    | |   |    | ||  |   |      ||         | |         ||   |  |    |      | ||   |   | |     ||  |   ||     | ||      | |      || |  |     |    |   |          | ||  |     |    ||       |   ||     ||  | ||   || |         |   |     |   |  ||    |      |  |  |  |        |||||   |    ||    |    |    |   |    |  |  | |    |   | |    |    |    |       |  ||  ||         | || |     |      |    |  |   |     |     |   | |  | |  |  || |   |      |   |      ||    ||    |  ||  | |   ||        |        ||  |    | \n",
      "---N-EFA----VD-----K--SKRG-L--IHVP--KDLPLKT---K-VLD-MS-QN---YI------------AELQV--S--------DMSF--L-S-E-L-TVL----RLSHNRIQL--LDLS-VFK------F-N-QDL---EY--L----DL-SH---NQLQKI--SCHPIV-----SF----R-HLD---L-SF------NDFKA--L----PICK-EFGNL-S-QLN---FL-GL-SAMKLQ--KLDLL-PI------AHLHLS---YIL-L---DL------RNYYIKE----NET-ESLQI-LNAK-T---LHLV--FHPT--SLFAIQV----N-IS-VNTLGCL-QL-----T-N---IKLN----D-D-NCQVFI---K-FLSE-LT--RGSTLLNFTLNHI-ETTWKCLVRVFQFLWPKPVEYL-NIY-NLTIIESIREEDFT-YSKTTLKALTIE----HIT--N--QVF-----LFSQ----TAL-Y---TVFSEMN-IM----ML--TIS---DTPFIHMLCPHA-PS--T--FKFL-NFTQN---VFTDS-IFEK---CSTLVKLETLILQKNGLKDLFKVGLMT--KDMPSLEILDVS-WNS--LES--GRHKE-NC-TW-VE----SIVVLNLSSNMLT---DSV--FRC----L-PPRI--KV----L-DL-HS-N----K---IKS--VPKQVVKL--EAL--QELNVAFNSLTDLP--GCGSFSSLSVL-I--IDHNSVSHPSA---DFFQS--CQ--KMRS----IK---A-G-DNPF-QC-TCE-L-REF-VKNIDQVSSE------VL--EGWP--DSY-KCD-YPES--YRG--------S-PLKDFHMSELSC--NITL\n",
      "  Score=237\n",
      "\n"
     ]
    }
   ],
   "source": [
    "alignments = pairwise2.align.globalxx(TLR2EC.seq, TLR6EC.seq)\n",
    "\n",
    "print(format_alignment(*alignments[0]))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reading in the fasta files for the transmembrane domain of TLR 2, TLR 4 and TLR 6.\n",
    "\n",
    "This region contains hydrophobic regions of the 3 TLRs that is embedded in the cells membrane\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "TLR2TM = SeqIO.read('TLR2TM.fasta','fasta')\n",
    "\n",
    "TLR4TM = SeqIO.read('TLR4TM.fasta','fasta')\n",
    "\n",
    "TLR6TM = SeqIO.read('TLR6TM.fasta','fasta')\n",
    "#SeqIO.read reads in the fasta file. As inputs, it takes the name of the file and the format seperated by a comma"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Performing pairwise sequence alignments: TLR 2 and TLR 4 vs TLR 2 and TLR 6 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AL----VSGMCCALFLLI-L--LTGV------LC--\n",
      "      ||           |  |  |      |   \n",
      "--TIIGVS----------VLSVL--VVSVVAVL-VY\n",
      "  Score=6\n",
      "\n"
     ]
    }
   ],
   "source": [
    "alignments = pairwise2.align.globalxx(TLR2TM.seq, TLR4TM.seq)\n",
    "\n",
    "print(format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AL-VS--GMCCALFLLI--L-L--TGV--LC-\n",
      " | |   |   |       | |  | |  || \n",
      "-LIV-TIG---A-----TMLVLAVT-VTSLCI\n",
      "  Score=10\n",
      "\n"
     ]
    }
   ],
   "source": [
    "alignments = pairwise2.align.globalxx(TLR2TM.seq, TLR6TM.seq)\n",
    "\n",
    "print(format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reading in the fasta files for the intracellular domains of TLR 2, TLR 4 and TLR 6\n",
    "This is the region present inside the cell which is responsible for the activation of the downstream signalling cascade after receptor activation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "TLR2CY = SeqIO.read('TLR2CY.fasta','fasta')\n",
    "\n",
    "TLR4CY = SeqIO.read('TLR4CY.fasta','fasta')\n",
    "\n",
    "TLR6CY = SeqIO.read('TLR6CY.fasta','fasta')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Performing pairwise sequence alignments: TLR 2 and TLR 4 vs TLR 2 and TLR 6 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HR-F--HGLWYMKMMWAWLQ-AKRKPR---KAPS--R--NICYDAFVS-YS---ERDAYWVE-N-LMVQE--LENFN---PPFK-LCLHK-RDFIPGKWI--ID--NIIDSI-E---KSH-KTVF-VL--SEN--FVK--SE-WCK---YELD------F-SHF-RLFDENNDAA-ILILLEP-IE---K--KAIP----Q-----RFCKL--RKIMNTKTYLEWPMDEAQRE----G---FWVN--LRA-AI----KS-------------------\n",
      "   |  | |     |   |  |        |     |  || |||||  ||   | |  ||  | | |    ||      |||  ||||  ||||||     |   | |  | |   ||  | |  |   |    |    |  ||    ||        | |   |        | | |     |    |  |       |     |   |  |   |  |||||       |    |   ||    ||  |     ||                   \n",
      "--KFYFH-L-----M---L-LA-----GCIK---YGRGENI-YDAFV-IYSSQDE-D--WV-RNEL-V--KNLE---EGVPPF-QLCLH-YRDFIPG---VAI-AAN-I--IHEGFHKS-RK-V-IV-VVS--QHF--IQS-RWC-IFEYE--IAQTWQFLS--SR--------AGI-I----FI-VLQKVEK---TLLRQQVELYR---LLSR---N--TYLEW-------EDSVLGRHIFW--RRLR-KA-LLDGKSWNPEGTVGTGCNWQEATSI\n",
      "  Score=84\n",
      "\n"
     ]
    }
   ],
   "source": [
    "alignments = pairwise2.align.globalxx(TLR2CY.seq, TLR4CY.seq)\n",
    "\n",
    "print(format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HRFHG-L---WYMKM--MWA---WL-QAK-RKPRK-APSRNICYD------------AFV-SYSER-D-AYWV--ENLMVQE--LENFNPPFKL-----CLHK-RD-FI-PGKWI-ID--NIIDS--IEKSH-KTV--FVLSE-NFVK-SEWCK-YELD-FS-HFR--LFD-E--NNDAA-ILILLEPIEKKAIPQRFC-----K---LRKI--MNTK--TYLE-WPMD-EAQ---REG-FWV-NL-RAAI---K-----------S\n",
      "      |   ||     |     |  |   |  |  |  |||               ||  ||||  | | ||  | | |    ||      |      |||  |  |  |||   |   |||    ||||  |    ||||  |||  ||||  |||  |  |    ||  |  ||    |||||||     |||        |   | |   | |   |||  ||   |     | | ||  |  |||    |           |\n",
      "-----YLDLPWY---LRM--VCQW-TQ--TR--R-RA--RNI---PLEELQRNLQFHAF-ISYSE-HDSA-WVKSE-L-V--PYLE------K-EDIQICLH-ER-NF-VPGK--SI-VENII--NCIEKS-YK--SIFVLS-PNFV-QSEWC-HYEL-YF-AH--HNLF-HEGSNN---LILILLEP-----IPQ---NSIPNKYHKL-K-ALM-T-QRTYL-QWP--KE--KSKR-GLFW-AN-IRAA-FNMKLTLVTENNDVKS\n",
      "  Score=97\n",
      "\n"
     ]
    }
   ],
   "source": [
    "alignments = pairwise2.align.globalxx(TLR2CY.seq, TLR6CY.seq)\n",
    "\n",
    "print(format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extracellular domain"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### The score for an alignment between TLR2 and TLR6's Extracellular domain is 250"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### The score for an alignment between TLR2 and TLR6's Extracellular domain is 237"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Transmembrane domain"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### The score for an alignment between TLR2 and TLR4's Transmembrane domain is 6\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### The score for an alignment between TLR2 and TLR6's Transmembrane domain is 10\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Cytoplasmic domain"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### The score for an alignment between TLR2 and TLR4's Cytoplasmic domain is 84"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### The score for an alignment between TLR2 and TLR6's Cytoplasmic domain is 97"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Vizually representing the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYgAAAEKCAYAAAAIO8L1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAHuxJREFUeJzt3XucHFWZ//HPNyEkhASQXCAkYCIGlGuA4eKiAgoqLAoiclEJCBLU4EqAVZZ1MSooKoIrCAjKEhG5/JZLWOXHPdyUQCYhQDBcQgjLQCQhRC7GQBKe/aPOJJVJTU9lMj3dM/N9v1796qrTp6qe7pqep+ucqlOKCMzMzFrqVesAzMysPjlBmJlZIScIMzMr5ARhZmaFnCDMzKyQE4SZmRVygjAzs0JOEGZmVsgJwszMCq1X6wDWxeDBg2PkyJG1DsPMrEuZPn36qxExpK16XTpBjBw5ksbGxlqHYWbWpUh6oUw9NzGZmVkhJwgzMyvkBGFmZoW6dB+EmdnaWLZsGU1NTSxdurTWoXSKfv36MWLECPr06dOu5Z0gzKzHaGpqYuDAgYwcORJJtQ6nqiKCRYsW0dTUxKhRo9q1jqo1MUnaUtIUSbMlPSnpm6l8oqSXJM1Mj4Nyy/ybpDmSnpb0yWrFZmY909KlSxk0aFC3Tw4Akhg0aNA6HS1V8whiOXBaRMyQNBCYLunO9NoFEXFevrKk7YCjgO2BLYC7JG0TESuqGKOZ9TA9ITk0W9f3WrUjiIiYHxEz0vSbwGxgeIVFDgGujYi3I+J5YA6wR7XiMzOzyjrlLCZJI4FdgIdT0cmSHpd0haT3pLLhwIu5xZqonFDMzNaJ1LGPtixatIgxY8YwZswYNt98c4YPH75yvn///mvUnzhx4so62223Hddcc80adc477zwk8eqrr3bER7KaqicISQOAG4BTIuIN4BJga2AMMB/4WXPVgsWjYH3jJDVKaly4cOE6xFVfDzPr/gYNGsTMmTOZOXMmX/3qV5kwYcLK+V69iv8dN9eZPHkyJ510EsuWLVv52osvvsidd97JVlttVZV4q5ogJPUhSw5XR8SNABHxSkSsiIh3gctZ1YzUBGyZW3wE8HLLdUbEZRHREBENQ4a0OZSImVm3MHr0aPr378/ixYtXlk2YMIGf/OQnVetXqeZZTAJ+A8yOiPNz5cNy1T4LzErTtwBHSeoraRQwGnikWvGZmXUlM2bMYPTo0QwdOhSAW265heHDh7PzzjtXbZvVPItpb+AY4AlJM1PZmcDRksaQNR/NA04CiIgnJV0P/IXsDKjxPoPJzHq6Cy64gMsvv5y5c+dy2223AbBkyRLOOecc7rjjjqpuu2oJIiIepLhf4dYKy5wDnFOtmMzMupoJEyZw+umnc+ONNzJ27Fiee+45nnvuOZ5//vmVRw9NTU3suuuuPPLII2y++eYdtm2PxWRm1gUcdthhNDQ0MGnSJHbccUcWLFjAvHnzmDdvHiNGjGDGjBkdmhzACcLMerCIjn2siyVLljBixIiVj/PPP3+NOmeddRbnn38+77777rptrCSPxWRmVgMTJ05cbb7MP/3ddtuNp59+eo3yefPmdVBUq/MRhJmZFXKCMDOzQm5iqhf1dDn1ujammlm34CMIMzMr5ARhZmaFnCDMzKyQE4SZ9VydPCxzRw/3feGFF7Ltttuy/fbb861vfavDPpZm7qQ2M+skzcN9Q/bPf8CAAZx++ukADBgwoHCZ5qE2nn32WXbbbTcOP/xw+vTpw5QpU5g8eTKPP/44ffv2ZcGCBR0er48gzMy6gJbDfV9yySWcccYZ9O3bF2DlKK8dyQnCzKwLaDnc9zPPPMMDDzzAnnvuyT777MO0adM6fJtuYjIzq2NFw30DLF++nMWLFzN16lSmTZvGEUccwdy5czv05kE+gjAzq2MTJkzg6aef5rrrrmPs2LEsXboUgBEjRnDYYYchiT322INevXp1+H2pnSDMzLqA/HDfAIceeij33HMPkDU3vfPOOwwePLhDt+kmJjPruepoWJnm4b6bnXrqqWvUOeuss/jCF77AiSeeyPHHH8/xxx/PDjvswPrrr8+kSZM6/N7Uijr6gNZWQ0NDNDY2tmvZehr6CCAKb75XI134b8KsktmzZ/PBD36w1mF0qqL3LGl6RDS0taybmMzMrJAThJmZFXKCMLMepSs3q6+tdX2vThBm1mP069ePRYsW9YgkEREsWrSIfv36tXsdPovJzHqMESNG0NTUxMKFC2sdSqfo16/famdGrS0nCDPrMfr06cOoUaNqHUaX4SYmMzMr5ARhZmaFWk0QkraUdK2kBySdKalP7rWbOyc8MzOrlUpHEFcA9wLfAIYB90kalF57b5XjMjOzGqvUST0kIi5N09+Q9CXgfkmfAbr/OWJmZj1cpQTRR1K/iFgKEBG/k/RX4HZgw06JzszMaqZSE9OvgT3zBRFxF/B5YFY1gzIzs9pr9QgiIi5opfxR4ICqRWRmZnWhaqe5prOgpkiaLelJSd9M5ZtKulPSs+n5Palckn4haY6kxyXtWq3YzMysbdW8DmI5cFpEfBDYCxgvaTvgDODuiBgN3J3mAQ4ERqfHOOCSKsZmZmZtqFqCiIj5ETEjTb8JzAaGA4cAk1K1ScChafoQ4LeRmQpsImlYteIzM7PKKiYISb0lDc7Nry9pnKTZa7MRSSOBXYCHgc0iYj5kSQQYmqoNB17MLdaUyszMrAYqXUl9FPAa8Lik+yTtB8wlawr6YtkNSBoA3ACcEhFvVKpaULbG9RYpQTVKauwpIzKamdVCpesgvgPsFhFzUofxQ8BREXFT2ZWn4TluAK6OiBtT8SuShkXE/NSEtCCVNwFb5hYfAbzccp0RcRlwGWT3pC4bi5mZrZ1KTUzvRMQcgNSX8PxaJgcBvwFmR8T5uZduAY5N08cCk3PlY9PZTHsBrzc3RZmZWeerdAQxVNKpufkB+fkW//SL7A0cAzwhaWYqOxM4F7he0gnA/5JdeAdwK3AQMAdYAny59LswM7MOVylBXA4MbGW+zaadiHiQ4n4FgI8X1A9gfFvrNTOzzlHpSurvtfaapFOqE46ZmdWL9l4HcWrbVczMrCtrb4JorenIzMy6ifYmCJ9eambWzbXaByHpTYoTgYANqhaRmZnVhUpnMW0aEcs6LRIzM6srlZqYHu60KMzMrO5UShDuiDYz68EqNTENaXEl9WpKXEltZmZdWKUE0RsYgI8kzMx6pEoJYn5EfL/TIjEzs7riPggzMytUKUGsMaCemZn1HK0miIh4rTMDMTOz+tLeoTbMzKybc4IwM7NCbSYISYdJelbS65LekPSmpDc6IzgzM6udSqe5NvsJ8OmImF3tYMzMrH6UaWJ6xcnBzKznqTTc92FpslHSdcDNwNvNr0fEjVWOzczMaqhSE9Onc9NLgE/k5gNwgjAz68ZaTRAR8WUASXtHxJ/yr0nau9qBmZlZbZXpg7iwZJmZmXUjlfogPgT8E2sO+70R2UivZmbWjVXqg1ifbLjv9YCBufI3gMOrGZSZmdVepT6I+4D7JF0ZES90YkxmZlYHKjUx/TwiTgEukhQtX4+Iz1Q1MjMzq6lKTUxXpefzOiMQMzOrL5WamKanyd7A1IhY0jkhmZlZPSgzFtNxwKWSFgEPpMeDEbG4moGZmVlttZkgImIsgKQtyM5e+iWwRZllzcys62rzn7ykLwEfAXYEXgUuIjuKMDOzbqzMUcDPgeeAS4EpETGvqhGZmVldaHOojYgYDBwP9APOkfSIpKvaWAxJV0haIGlWrmyipJckzUyPg3Kv/ZukOZKelvTJdr4fMzPrIGXuKLcRsBXwXmAksDHwbol1Xwl8qqD8gogYkx63pm1sBxwFbJ+WuViSh/MwM6uhMk1MD+YeF0VEU5kVR8T9kkaWjOMQ4NqIeBt4XtIcYA/goZLLm5lZBytzFtNOHbzNkyWNBRqB09LpssOBqbk6TanMzMxqpMxw3x3pEmBrYAwwH/hZKldB3TWG9wCQNE5So6TGhQsXVidKMzPr3AQREa9ExIqIeBe4nKwZCbIjhi1zVUcAL7eyjssioiEiGoYMGVLdgM3MerBOTRCShuVmPws0n+F0C3CUpL6SRgGjgUc6MzYzM1tdmQvltiFrGtosInaQtBPwmYg4u43lrgH2BQZLagK+C+wraQxZ89E84CSAiHhS0vXAX4DlwPiIWNHud2VmZutMEYVN/asqSPcB/wr8KiJ2SWWzImKHToivooaGhmhsbGzXsirq9aihKOyGqZE2/ibMrGuTND0iGtqqV6aJqX9EtGzuWd6+sMzMrKsokyBelbQ16awiSYeTnYFkZmbdWJkL5cYDlwEfkPQS8DzwxapGZWZmNVcxQUjqBTRExP6SNgR6RcSbnROamZnVUsUmpnS9wslp+u9ODmZmPUeZPog7JZ0uaUtJmzY/qh6ZmZnVVJk+iOPT8/hcWQDv6/hwzMysXpQZrG9UZwRiZmb1pcyV1H2ArwEfTUX3kl00t6yKcZmZWY2VaWK6BOgDXJzmj0llX6lWUGZmVntlEsTuEbFzbv4eSY9VKyAzM6sPZc5iWpGupAZA0vsAD6RnZtbNlTmC+FdgiqS5ZDf2eS/w5apGZWZmNVfmLKa7JY0GtiVLEE+le0ebmVk31mYTk6TxwAYR8XhEPAb0l/T16odmZma1VKYP4sSI+FvzTEQsBk6sXkhmZlYPyiSIXtKq2+tI6g2sX72QzMysHpTppL4duF7SpWRDbHwVuK2qUZmZWc2VSRDfBsaRXU0t4A7g19UMyszMaq/MWUzvApdKugLYHngpInwdhJlZN9dqH4SkSyVtn6Y3BmYCvwUelXR0J8VnZmY1UqmT+iMR8WSa/jLwTETsCOwGfKvqkZmZWU1VShDv5KYPAG4GiIi/VjUiMzOrC5USxN8kHSxpF2Bv0plLktYDNuiM4MzMrHYqdVKfBPwC2Bw4JXfk8HHgj9UOzMzMaqvVBBERzwCfKii/nezaCDMz68bKXEltZmY9kBOEmZkVKjOa66gyZWZm1r2UOYK4oaDsvzs6EDMzqy+tdlJL+gDZ0BobSzos99JGQL9qB2ZmZrVV6TTXbYGDgU2AT+fK38T3gzAz6/YqneY6GZgs6UMR8dDarjgN7ncwsCAidkhlmwLXASOBecAREbE43W/iP4GDgCXAcRExY223aWZmHadMH8QcSWdKukzSFc2PEstdyZrXUZwB3B0Ro4G70zzAgcDo9BgHXFIqejMzq5oy94OYDDwA3AWUHuY7Iu6XNLJF8SHAvml6EnAv2f0mDgF+GxEBTJW0iaRhETG/7PbMzKxjlUkQ/SPi2x20vc2a/+lHxHxJQ1P5cODFXL2mVOYEYWZWI2WamP4g6aAqx6GCsiisKI2T1CipceHChVUOy8ys5yqTIL5JliSWSnpD0puS3mjn9l6RNAwgPS9I5U3Alrl6I4CXi1YQEZdFRENENAwZMqSdYZiZWVvaTBARMTAiekVEv4jYKM1v1M7t3QIcm6aPJevfaC4fq8xewOvufzCzuiPVz6MTlBlqQ5K+JOk/0vyWkvYosdw1wEPAtpKaJJ0AnAscIOlZspsQnZuq3wrMBeYAlwNfb9e7MTOzDlOmk/pi4F3gY8APgLeAXwK7V1ooIlq7b/XHC+oGML5ELGZm1knKJIg9I2JXSY8CpAvb1q9yXGZmVmNlOqmXSepNOqtI0hCyIwozM+vGyiSIXwA3AUMlnQM8CPywqlGZmVnNtdnEFBFXS5pO1ncg4NCImF31yMzMrKbK9EEAvEI23MZ6wAaSdvVgemZm3VubCULSD4DjgOdYdXVzkJ3VZGZm3VSZI4gjgK0j4p1qB2NmZvWjTCf1LLKbBpmZWQ9S5gjiR8CjkmYBbzcXRsRnqhaVmZnVXJkEMQn4MfAEvv7BzDpZJw07VErhENPdWJkE8WpE/KLqkZiZWV0pkyCmS/oR2Yir+SYmn+ZqZtaNlUkQu6TnvXJlPs3VzKybK3Ml9X6dEYiZmdWXMhfK9QU+B4zM14+I71cvLDMzq7UyTUyTgdeB6eT6IMzMrHsrkyBGRMSnqh6JmZnVlTJXUv9Z0o5Vj8TMzOpKmSOIDwPHSXqerIlJZHcJ3amqkZmZWU2VSRAHVj0KMzOrO2USxJsly8zMrBsp0wcxA1gIPAM8m6aflzRD0m7VDM7MzGqnTIK4DTgoIgZHxCCyJqfrga8DF1czODMzq50yCaIhIm5vnomIO4CPRsRUoG/VIjMzs5oq0wfxmqRvA9em+SOBxZJ64+G/zcy6rTJHEF8ARgA3k11VvVUq6012O1IzM+uGygzW9yrwjVZentOx4ZiZWb1oNUFI+nlEnCLpfyi4kZJvOWpm1r1VOoK4Kj2f1xmBmJlZfWk1QUTE9PR8X+eFY2Zm9aJSE9MTFN+j22MxmZn1AJWamA7utCjMzKzuVGpieqGoXNLeZKe5jq9WUGZmVntlLpRD0hiypHAE8Dxw47psVNI8sgH/VgDLI6JB0qbAdWS3Np0HHBERi9dlO2Zm1n6tXignaRtJZ0maDVwEvAgoIvaLiAs7YNv7RcSYiGhI82cAd0fEaODuNG9mZjVS6Urqp4CPA5+OiA+npLCiirEcAkxK05OAQ6u4LTMza0OlBPE54K/AFEmXS/o42RlMHSGAOyRNlzQulW0WEfMB0vPQogUljZPUKKlx4cKFHRSOmZm11GqCiIibIuJI4APAvcAEYDNJl0j6xDpud++I2JVs6PDxkj5adsGIuCwiGiKiYciQIesYhpmZtabNwfoi4u8RcXVEHEw2aN9M1rF/ICJeTs8LgJuAPYBXJA0DSM8L1mUbZma2bsqM5rpSRLwWEb+KiI+1d4OSNpQ0sHka+AQwC7gFODZVO5Zs5FgzM6uRUqe5drDNgJskNW//9xFxm6RpwPWSTgD+F/h8DWIzM7Ok0xNERMwFdi4oX0R21pSZmdWBtWpiMjOznsMJwszMCjlBmJlZIScIMzMr5ARhZmaFnCDMzKyQE4SZmRVygjAzs0JOEGZmVsgJwszMCjlBmJlZIScIMzMr5ARhZmaFnCDMzKyQE4SZmRVygjAzs0JOEGZmVsgJwszMCjlBmJlZIScIMzMr5ARhZmaFnCDMzKyQE4SZmRVygjAzs0JOEGZmVsgJwszMCjlBmJlZIScIMzMrtF6tAzDrtqRaR7BKRK0jsC7IRxBmZlbICcLMzAq5icm6jXpq0QFwo451dXV3BCHpU5KeljRH0hm1jsfMrKeqqwQhqTfwS+BAYDvgaEnb1TYqM7Oeqa4SBLAHMCci5kbEO8C1wCE1jsnMrEeqtwQxHHgxN9+UyszMrJPVWyd1UTfjan19ksYB49LsW5KernpUnaCD+lcHA6+u81rqrbe3i/I+7X660T59b5lK9ZYgmoAtc/MjgJfzFSLiMuCyzgyqq5DUGBENtY7DOo73affTlfZpvTUxTQNGSxolaX3gKOCWGsdkZtYj1dURREQsl3QycDvQG7giIp6scVhmZj1SXSUIgIi4Fbi11nF0UW566368T7ufLrNPFR7Ey8zMCtRbH4SZmdUJJ4gqkLRC0szco+KQIZLO7ISYRkqalab3lfSHEsu8Ve24aknSoNw++qukl3Lz69c6vvaS1CRpk1rH0RVI2lzStZKek/QXSbdK2qaVumMkHbQO21r5HawGSQ2SftGR66y7Pohu4h8RMWYt6p8J/LBloSSRNQO+22GRVUlXirVZRCwCxgBImgi8FRHn5et0xfdVhqT1ImJ5reOopbRvbwImRcRRqWwMsBnwTMEiY4AG6rSPNCIagcaOXKePIDqJpI3TIITbpvlrJJ0o6Vxgg/Sr9er0K2O2pIuBGcCWki6R1CjpSUnfy61zd0l/lvSYpEckDZTUW9JPJU2T9Likk9qIa6Kk03PzsySNbFFngKS7Jc2Q9ISkQ1L5GrF20MdVU5Lenz6HS8ne1zBJl+X2wVm5uk3pM3w0fd7bpPKPpf0yM31uG0raX9IUSf8t6VlJZ0sam9tXI9Oym0m6MW3vEUl7pfKzJV0p6Q5J8yQdKulnKdY/Ssr/4DsjLfuwpPel5X+X6k8BfihpL0kPpdj/JGl0qveVFOPtKc4f5d7vgWmZGZKuk7RhdfdGVe0HLIuIS5sLImImMK75bxwgfS8/A3wfODLt0yMlbSrp5rTvpkraKdWfKOkqSfekz+/ElhtO350H0uc4Q9I/pfJhku5P25gl6SOp/C1JP5Y0XdJdkvaQdK+kuSm21VoG0nf2v9L39XFJn2vXJxQRfnTwA1gBzMw9jkzlBwAPkV3fcVuu/lu56ZHAu8BeubJN03Nv4F5gJ2B9YC6we3ptI7IjwnHAd1JZX7JfFKPSemel8n2BP6TpicDpuW3NAkbm40rr3ShNDwbmkF1UukasXfWR/xyA96f3tXvBPlgPeADYLs03AV9L0/8CXJqm/z+wZ5oekPbd/sBrZL9Q+wF/Bc5KdU4DzkvT1zV/pi3229nAfSmG3YAlwAHptf8BDs7F9O00fTxwc5r+HXAz0CvNbwz0TtOfAq5L018BngUGAhuQDX+zBTA0bb9/qvfvwJm13nfrsM//BbigoHyf3Ge2MfB8+syPAy7K1bsQ+G6a/hgwM/e39Fj67AbnPr/8vuwP9EvTo4HG3N/Bv6fp3sDANB3AgWn6JuAOoA+wc267+7Lqe/1j4Oe5WN/Tns/ITUzVUdjEFBF3Svo82Yi1O1dY/oWImJqbP0LZECPrAcPIRroNYH5ETEvrfgNA0ieAnSQdnpbdmOwPsOiQuSyR/eL8KNk/zuFk/+SKYu0unmv+bJOjJZ1Atg+2INsHf0mv3ZiepwPNbdR/An4u6ffADRHxlrKhER6OiFcAJM0lu+YH4AngQ2l6f2BbrRpK4T2SNkjTt0Z2vdATkP1N5ZYfmYv3mvR8NXBurvz/xarmsk2A30rauuD93xURb6Y4nwK2AjZP7/vPKbb1gQcLlu3SIuI+Sb+UNBQ4jGz/LdeaQ1t8GPhcWuYeZX1aG6fXJkfEP4B/pCO2Pch+LDbrA1ykrElrBdDc7zENuEJSH7Ik1bzMO8BtafoJ4O2IWJb+DkYWvI39yX6INr+nxWv3KWScIDqRpF7AB4F/AJuS/dIr8vfcMqOA08l+zS6WdCXZr09RfE8aAd+IiNtXK2zRbJSznNWbGvsV1PkiMATYLf1RzsvV+3tB/e4gvw9GA98E9oiIv0n6Hat/Tm+n5xWk71REnC3pFuCfgWmS9m1RF7Jk+3Zuuvn7qLStd/IBpX9Q+fr51/PLQ+v3K8rvr3OA2yPiYknvZ9U/oJZxNr8vkR35HtPKuruaJ4HDW3ntKrK/+6PIjsKKVBo7ruXn33J+AvAK2Q/FXsBSgIi4P/0Q+2fgKkk/jYjfkjWFNa9j5d9NRLzbomkxH9s6X8PgPojONQGYDRzNql8JAMty0y1tRPalfl3SZmT3ygB4CthC0u4Ayvof1iP7Rfq15vVJ2qaNduJ5wK6p7q5kzVEtbQwsSMlhP0oO9NWNbAS8CbwhaRjwybYWkLR1RDweET8CHgW2XYvt3QWMz61rbU54aHZkej6a7GimyMbAS2n6uBLr/DOwT65PY8Pmfosu6h6gb76PQFm/3j7AlcApALFqNIc3yZrdmt1PlkRIPwBebT6SBw6R1E/SILKmn/zRKGSf/fx0NHcMWXMSkt5L9l27HPgN6bvZDncAJ+fe13vasxIniOpo7nRufpyrrPPyK8BpEfEA2R/Xd1L9y4DHJV3dckUR8RjZP5gngStIX/b06/JI4EJJjwF3kv2q/TVZ08cMZafU/YrKR4o3AJtKmgl8jeKmqKuBBkmNZF+Ip9bis+gOZpB9prOAy2n9H27e6amT8XHgb2Rf2LLGA3unzsW/AGt0cpbQX9IjZPv0tFbq/Bj4qaQy74fUNHYCcF36m/szq5pGupz0i/yzwAHKTnN9kqz/4OX0XmcD/5VbZAqwXXMndarbkPbxucCxubqPAH8EpgI/iIjVBh0FLgaOlTSV7DNsPrLbF5gp6VGy5qv/bOfbO5usaXJW2lf7tWclvpLazKwFSf3J2vp3jYjX13LZiRScMt0V+QjCzCxH0v5kR8kXrm1y6G58BGFmZoV8BGFmZoWcIMzMrJAThJmZFXKCMMvRqpF4n1Q2ltKp6QLHam7zq5LGVnMbZu3hTmqzHElvRcSAND0U+D3wp4j4bm0jM+t8ThBmOfkEkebfR3YV7GCywQ8vIRvyeTlwakRMkXQccCjZ1bA7AD8jG6foGLIhEQ6KiNfSFbvj0mtzgGMiYkn+vHlJ9wIPk13YtAlwQrqw0qzTuYnJrIKImEv2PRlKGv4iInYkG8JikqTmMZl2AL5ANijbOcCSiNiFbPTe5uajGyNi94jYmewq3RNa2ex6EbEH2VAPPnKxmnGCMGtb86BsHyYbxI2IeAp4gVVDTUyJiDcjYiHwOtnw27D6KKs7KLsHwBNkQ5Zs38r28qPDjmyljlnVOUGYVZCamFYACygevbNZmVFarwROTkcg36N45Nz8ulaODmtWC04QZq2QNAS4lOwmMcHqo3duQ3aPhKfXYpUDgflppN0vdnC4Zh3Ov07MVrdBGtm2D1lH9FXA+em1i4FLUxPRcuC4iHi74EYyrfkPsg7oF8iangZWrm5WWz6LyczMCrmJyczMCjlBmJlZIScIMzMr5ARhZmaFnCDMzKyQE4SZmRVygjAzs0JOEGZmVuj/AFXu/ky4kQasAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "x = np.arange(3) \n",
    "TLR4 = [250, 6, 83] \n",
    "TLR6 = [237, 10, 97 ] \n",
    "#creating a numpy array for our pairwise sequence alignment scores  \n",
    "\n",
    "width = 0.3\n",
    "#determing width of the bars\n",
    "\n",
    "plot.bar(x-0.3, TLR4, width, color='blue') \n",
    "plot.bar(x, TLR6, width, color='red') \n",
    "#Plotting the alignment scores and color coding the two groups\n",
    "\n",
    "plot.xticks(x, ['Extracellular', 'Transmembrane', 'Cytoplasmic']) \n",
    "plot.xlabel(\"Domain\") \n",
    "plot.ylabel(\"Alignment Score with TLR 2\") \n",
    "plot.legend([ \"TLR4\", \"TLR6\"]) \n",
    "#Creating plot legend and naming axis\n",
    "plot.show() \n",
    "#Command to vizualize the plot\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Conclusion\n",
    "\n",
    "#### Sequence allignment performed for TLR2 vs TLR4 for the extracellular domain has a higher score value as compared to the sequence alignment performed for TLR2 vs TLR6. This concludes that TLR2 extracellar gene sequence that is responsible for recognizing and binding ligands is more structurally and functionally similar to the TLR4 gene sequence as compared to the TLR6 gene sequence even though TLR2 and TLR6 dimerize to activate downstream signalling. This implies that TLR 2 and 4 show a greater overlap in the structural motifs their ligand binding domains recognize.\n",
    "\n",
    "#### Similarly, as hypothesized, the TLR2 gene sequences that are responsible for dimerization and downstream signalling, the hydrophobic transmembrane and intracellular domains, are more similar to the TLR6 gene sequences as compared to the TLR4 gene sequence even though TLR 4 most recently evolved from TLR 2. This implies that TLR2 and 6 share a more similar mechanism of downstream signal activation as compared to TLR2 and TLR4.\n",
    "\n",
    "#### In conclusion, our hypothesis was true ie we fail to reject our hypothesis\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### References and Sources"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.uniprot.org/uniprot/O60603\n",
    "https://www.uniprot.org/uniprot/O00206\n",
    "https://www.uniprot.org/uniprot/Q9Y2C9\n",
    "https://www.rcsb.org/structure/6NIG\n",
    "https://www.rcsb.org/structure/3FXI\n",
    "https://www.frontiersin.org/articles/10.3389/fimmu.2014.00316/full"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
